File: src/test/java/org/example/kalkulationsprogramm/mapper/ProduktkategorieMapperTest.java
package org.example.kalkulationsprogramm.mapper;

import org.example.kalkulationsprogramm.domain.Produktkategorie;
import org.example.kalkulationsprogramm.domain.Verrechnungseinheit;
import org.example.kalkulationsprogramm.dto.Produktkategroie.ProduktkategorieResponseDto;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

class ProduktkategorieMapperTest {

    private final ProduktkategorieMapper mapper = new ProduktkategorieMapper();

    @Test
    void mapptAlleFelder() {
        Produktkategorie kategorie = new Produktkategorie();
        kategorie.setId(1L);
        kategorie.setBezeichnung("Stahl");
        kategorie.setVerrechnungseinheit(Verrechnungseinheit.KILOGRAMM);
        kategorie.setBeschreibung("Stahlprodukte");
        kategorie.setBildUrl("stahl.png");

        ProduktkategorieResponseDto dto = mapper.toProduktkategorieResponseDto(kategorie);

        assertThat(dto.getId()).isEqualTo(1L);
        assertThat(dto.getBezeichnung()).isEqualTo("Stahl");
        assertThat(dto.getVerrechnungseinheit()).isEqualTo(Verrechnungseinheit.KILOGRAMM);
        assertThat(dto.getBeschreibung()).isEqualTo("Stahlprodukte");
        assertThat(dto.getBildUrl()).isEqualTo("/api/images/stahl.png");
        assertThat(dto.isLeaf()).isTrue();
    }

    @Test
    void gibtNullZurueckBeiNullKategorie() {
        assertThat(mapper.toProduktkategorieResponseDto(null)).isNull();
    }

    @Test
    void setzLeafAufFalseBeiVorhandenenUnterkategorien() {
        Produktkategorie kind = new Produktkategorie();
        kind.setId(2L);
        kind.setBezeichnung("Unterkategorie");

        Produktkategorie kategorie = new Produktkategorie();
        kategorie.setId(1L);
        kategorie.setBezeichnung("Eltern");
        kategorie.setVerrechnungseinheit(Verrechnungseinheit.STUECK);
        kategorie.setUnterkategorien(List.of(kind));

        ProduktkategorieResponseDto dto = mapper.toProduktkategorieResponseDto(kategorie);

        assertThat(dto.isLeaf()).isFalse();
    }

    @Test
    void bauePfadMitHierarchie() {
        Produktkategorie opa = new Produktkategorie();
        opa.setId(1L);
        opa.setBezeichnung("Werkstoffe");
        opa.setVerrechnungseinheit(Verrechnungseinheit.KILOGRAMM);

        Produktkategorie vater = new Produktkategorie();
        vater.setId(2L);
        vater.setBezeichnung("Metalle");
        vater.setVerrechnungseinheit(Verrechnungseinheit.KILOGRAMM);
        vater.setUebergeordneteKategorie(opa);

        Produktkategorie kind = new Produktkategorie();
        kind.setId(3L);
        kind.setBezeichnung("Stahl");
        kind.setVerrechnungseinheit(Verrechnungseinheit.KILOGRAMM);
        kind.setUebergeordneteKategorie(vater);

        ProduktkategorieResponseDto dto = mapper.toProduktkategorieResponseDto(kind);

        assertThat(dto.getPfad()).isEqualTo("Werkstoffe > Metalle > Stahl");
    }

    @Test
    void bauePfadOhneElternkategorie() {
        Produktkategorie kategorie = new Produktkategorie();
        kategorie.setId(1L);
        kategorie.setBezeichnung("Einzelne Kategorie");
        kategorie.setVerrechnungseinheit(Verrechnungseinheit.STUECK);

        ProduktkategorieResponseDto dto = mapper.toProduktkategorieResponseDto(kategorie);

        assertThat(dto.getPfad()).isEqualTo("Einzelne Kategorie");
    }

    @Test
    void laeBildUrlMitSchraegstrichUnveraendert() {
        Produktkategorie kategorie = new Produktkategorie();
        kategorie.setId(1L);
        kategorie.setBezeichnung("Test");
        kategorie.setVerrechnungseinheit(Verrechnungseinheit.STUECK);
        kategorie.setBildUrl("/api/images/vorhandenes-bild.png");

        ProduktkategorieResponseDto dto = mapper.toProduktkategorieResponseDto(kategorie);

        assertThat(dto.getBildUrl()).isEqualTo("/api/images/vorhandenes-bild.png");
    }

    @Test
    void behandeltNullBildUrl() {
        Produktkategorie kategorie = new Produktkategorie();
        kategorie.setId(1L);
        kategorie.setBezeichnung("Test");
        kategorie.setVerrechnungseinheit(Verrechnungseinheit.STUECK);
        kategorie.setBildUrl(null);

        ProduktkategorieResponseDto dto = mapper.toProduktkategorieResponseDto(kategorie);

        assertThat(dto.getBildUrl()).isNull();
    }

    @Test
    void behandeltLeereBildUrl() {
        Produktkategorie kategorie = new Produktkategorie();
        kategorie.setId(1L);
        kategorie.setBezeichnung("Test");
        kategorie.setVerrechnungseinheit(Verrechnungseinheit.STUECK);
        kategorie.setBildUrl("   ");

        ProduktkategorieResponseDto dto = mapper.toProduktkategorieResponseDto(kategorie);

        assertThat(dto.getBildUrl()).isEqualTo("   ");
    }
}

File: src/test/kotlin/org/example/kalkulationsprogramm/mapper/ProduktkategorieMapperTest.kt
package org.example.kalkulationsprogramm.mapper

import org.assertj.core.api.Assertions.assertThat
import org.example.kalkulationsprogramm.domain.Produktkategorie
import org.example.kalkulationsprogramm.domain.Verrechnungseinheit
import org.junit.jupiter.api.Test

class ProduktkategorieMapperTest {

    private val mapper = ProduktkategorieMapper()

    @Test
    fun mapptAlleFelder() {
        val kategorie = Produktkategorie().apply {
            id = 1L
            bezeichnung = "Stahl"
            verrechnungseinheit = Verrechnungseinheit.KILOGRAMM
            beschreibung = "Stahlprodukte"
            bildUrl = "stahl.png"
        }

        val dto = mapper.toProduktkategorieResponseDto(kategorie)!!

        assertThat(dto.id).isEqualTo(1L)
        assertThat(dto.bezeichnung).isEqualTo("Stahl")
        assertThat(dto.verrechnungseinheit).isEqualTo(Verrechnungseinheit.KILOGRAMM)
        assertThat(dto.beschreibung).isEqualTo("Stahlprodukte")
        assertThat(dto.bildUrl).isEqualTo("/api/images/stahl.png")
        assertThat(dto.isLeaf).isTrue()
    }

    @Test
    fun gibtNullZurueckBeiNullKategorie() {
        assertThat(mapper.toProduktkategorieResponseDto(null)).isNull()
    }

    @Test
    fun setzLeafAufFalseBeiVorhandenenUnterkategorien() {
        val kind = Produktkategorie().apply {
            id = 2L
            bezeichnung = "Unterkategorie"
        }
        val kategorie = Produktkategorie().apply {
            id = 1L
            bezeichnung = "Eltern"
            verrechnungseinheit = Verrechnungseinheit.STUECK
            unterkategorien = mutableListOf(kind)
        }

        val dto = mapper.toProduktkategorieResponseDto(kategorie)!!

        assertThat(dto.isLeaf).isFalse()
    }

    @Test
    fun bauePfadMitHierarchie() {
        val opa = Produktkategorie().apply {
            id = 1L
            bezeichnung = "Werkstoffe"
            verrechnungseinheit = Verrechnungseinheit.KILOGRAMM
        }
        val vater = Produktkategorie().apply {
            id = 2L
            bezeichnung = "Metalle"
            verrechnungseinheit = Verrechnungseinheit.KILOGRAMM
            uebergeordneteKategorie = opa
        }
        val kind = Produktkategorie().apply {
            id = 3L
            bezeichnung = "Stahl"
            verrechnungseinheit = Verrechnungseinheit.KILOGRAMM
            uebergeordneteKategorie = vater
        }

        val dto = mapper.toProduktkategorieResponseDto(kind)!!

        assertThat(dto.pfad).isEqualTo("Werkstoffe > Metalle > Stahl")
    }

    @Test
    fun bauePfadOhneElternkategorie() {
        val kategorie = Produktkategorie().apply {
            id = 1L
            bezeichnung = "Einzelne Kategorie"
            verrechnungseinheit = Verrechnungseinheit.STUECK
        }

        val dto = mapper.toProduktkategorieResponseDto(kategorie)!!

        assertThat(dto.pfad).isEqualTo("Einzelne Kategorie")
    }

    @Test
    fun laeBildUrlMitSchraegstrichUnveraendert() {
        val kategorie = Produktkategorie().apply {
            id = 1L
            bezeichnung = "Test"
            verrechnungseinheit = Verrechnungseinheit.STUECK
            bildUrl = "/api/images/vorhandenes-bild.png"
        }

        val dto = mapper.toProduktkategorieResponseDto(kategorie)!!

        assertThat(dto.bildUrl).isEqualTo("/api/images/vorhandenes-bild.png")
    }

    @Test
    fun behandeltNullBildUrl() {
        val kategorie = Produktkategorie().apply {
            id = 1L
            bezeichnung = "Test"
            verrechnungseinheit = Verrechnungseinheit.STUECK
            bildUrl = null
        }

        val dto = mapper.toProduktkategorieResponseDto(kategorie)!!

        assertThat(dto.bildUrl).isNull()
    }

    @Test
    fun behandeltLeereBildUrl() {
        val kategorie = Produktkategorie().apply {
            id = 1L
            bezeichnung = "Test"
            verrechnungseinheit = Verrechnungseinheit.STUECK
            bildUrl = "   "
        }

        val dto = mapper.toProduktkategorieResponseDto(kategorie)!!

        assertThat(dto.bildUrl).isEqualTo("   ")
    }
}
